File: panel/app/api/tickets.py
"""工单 API：创建/回复/关闭。"""
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from ..core.database import get_db
from ..core.deps import get_current_user
from ..core.errors import BizError, NotFoundError, success
from ..models import Ticket, User
from ..schemas.schemas import TicketCreate, TicketReply

router = APIRouter(prefix="/api/tickets", tags=["工单"])


@router.post("")
def create_ticket(
    payload: TicketCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """创建工单。"""
    ticket = Ticket(
        user_id=user.id,
        title=payload.title,
        content=payload.content,
        status="open",
    )
    db.add(ticket)
    db.commit()
    db.refresh(ticket)
    return success(_serialize(ticket), "工单已提交", 201)


@router.get("")
def list_tickets(
    page: int = 1,
    page_size: int = 20,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """工单列表。"""
    q = db.query(Ticket).filter(Ticket.user_id == user.id).order_by(Ticket.id.desc())
    total = q.count()
    items = q.offset((page - 1) * page_size).limit(page_size).all()
    return success(
        {
            "total": total,
            "page": page,
            "page_size": page_size,
            "items": [_serialize(t) for t in items],
        }
    )


@router.get("/{ticket_id}")
def ticket_detail(
    ticket_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """工单详情。"""
    ticket = _get_owned_ticket(db, user, ticket_id)
    return success(_serialize(ticket))


@router.post("/{ticket_id}/reply")
def reply_ticket(
    ticket_id: int,
    payload: TicketReply,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """回复工单（用户补充信息）。"""
    ticket = _get_owned_ticket(db, user, ticket_id)
    if ticket.status != "open":
        raise BizError(400, 5001, "工单已关闭，无法回复")
    ticket.content += f"\n\n【用户回复】\n{payload.content}"
    db.commit()
    return success(_serialize(ticket), "回复成功")


@router.post("/{ticket_id}/close")
def close_ticket(
    ticket_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """关闭工单。"""
    ticket = _get_owned_ticket(db, user, ticket_id)
    if ticket.status != "open":
        raise BizError(400, 5001, "工单已关闭")
    ticket.status = "closed"
    db.commit()
    return success(_serialize(ticket), "工单已关闭")


def _serialize(ticket: Ticket) -> dict:
    return {
        "id": ticket.id,
        "user_id": ticket.user_id,
        "title": ticket.title,
        "content": ticket.content,
        "status": ticket.status,
        "admin_reply": ticket.admin_reply,
        "created_at": ticket.created_at.isoformat() if ticket.created_at else None,
        "updated_at": ticket.updated_at.isoformat() if ticket.updated_at else None,
    }


def _get_owned_ticket(db: Session, user: User, ticket_id: int) -> Ticket:
    ticket = (
        db.query(Ticket)
        .filter(Ticket.id == ticket_id, Ticket.user_id == user.id)
        .first()
    )
    if ticket is None:
        raise NotFoundError(message="工单不存在")
    return ticket

File: panel-go/api/tickets.go
package api

import (
	"strconv"

	"github.com/gin-gonic/gin"

	"weavenet/panel/database"
	"weavenet/panel/middleware"
	"weavenet/panel/models"
	"weavenet/panel/utils"
)

// CreateTicket 创建工单。
func CreateTicket(c *gin.Context) {
	var p struct {
		Title   string `json:"title"`
		Content string `json:"content"`
	}
	if err := c.ShouldBindJSON(&p); err != nil || p.Title == "" || p.Content == "" {
		utils.Fail(c, utils.Biz(400, 0, "标题和内容不能为空"))
		return
	}
	user := middleware.GetUser(c)
	t := models.Ticket{UserID: user.ID, Title: p.Title, Content: p.Content, Status: "open"}
	if err := database.DB.Create(&t).Error; err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	utils.OK(c, serializeTicket(&t), "工单已提交")
}

// ListTickets 工单列表。
func ListTickets(c *gin.Context) {
	user := middleware.GetUser(c)
	db := database.DB
	page, _ := utils.AtoiSafe(c.DefaultQuery("page", "1"), 1)
	pageSize, _ := utils.AtoiSafe(c.DefaultQuery("page_size", "20"), 20)
	var total int64
	db.Model(&models.Ticket{}).Where("user_id = ?", user.ID).Count(&total)
	var list []models.Ticket
	db.Where("user_id = ?", user.ID).Order("id desc").Offset((page - 1) * pageSize).Limit(pageSize).Find(&list)
	items := make([]gin.H, 0, len(list))
	for _, t := range list {
		items = append(items, serializeTicket(&t))
	}
	utils.OK(c, gin.H{"total": total, "page": page, "page_size": pageSize, "items": items}, "ok")
}

// TicketDetail 工单详情。
func TicketDetail(c *gin.Context) {
	user := middleware.GetUser(c)
	t, be := getOwnedTicket(user.ID, c.Param("ticket_id"))
	if be != nil {
		utils.Fail(c, be)
		return
	}
	utils.OK(c, serializeTicket(t), "ok")
}

// ReplyTicket 回复工单（用户补充信息）。
func ReplyTicket(c *gin.Context) {
	var p struct {
		Content string `json:"content"`
	}
	if err := c.ShouldBindJSON(&p); err != nil || p.Content == "" {
		utils.Fail(c, utils.Biz(400, 0, "回复内容不能为空"))
		return
	}
	user := middleware.GetUser(c)
	t, be := getOwnedTicket(user.ID, c.Param("ticket_id"))
	if be != nil {
		utils.Fail(c, be)
		return
	}
	if t.Status != "open" {
		utils.Fail(c, utils.Biz(400, 5001, "工单已关闭，无法回复"))
		return
	}
	t.Content += "\n\n【用户回复】\n" + p.Content
	if err := database.DB.Save(t).Error; err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	utils.OK(c, serializeTicket(t), "回复成功")
}

// CloseTicket 关闭工单。
func CloseTicket(c *gin.Context) {
	user := middleware.GetUser(c)
	t, be := getOwnedTicket(user.ID, c.Param("ticket_id"))
	if be != nil {
		utils.Fail(c, be)
		return
	}
	if t.Status != "open" {
		utils.Fail(c, utils.Biz(400, 5001, "工单已关闭"))
		return
	}
	t.Status = "closed"
	if err := database.DB.Save(t).Error; err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	utils.OK(c, serializeTicket(t), "工单已关闭")
}

func serializeTicket(t *models.Ticket) gin.H {
	return gin.H{
		"id": t.ID, "user_id": t.UserID, "title": t.Title, "content": t.Content,
		"status": t.Status, "admin_reply": t.AdminReply,
		"created_at": utils.TimeFmtV(t.CreatedAt), "updated_at": utils.TimeFmtV(t.UpdatedAt),
	}
}

func getOwnedTicket(userID uint, idStr string) (*models.Ticket, *utils.BizError) {
	id, err := strconv.Atoi(idStr)
	if err != nil {
		return nil, utils.NotFound("工单不存在")
	}
	var t models.Ticket
	if err := database.DB.Where("id = ? AND user_id = ?", id, userID).First(&t).Error; err != nil {
		return nil, utils.NotFound("工单不存在")
	}
	return &t, nil
}
